src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java

```diff
@@ -223,6 +223,46 @@ private static ImmutableMap<URI, Map<String, List<String>>> getAuthHeaders(
     return res;
   }
 
+  private static ImmutableMap<URI, Map<String, List<String>>> getHeaders(
+    Map<String, Map<String, List<String>>> auth) throws RepositoryFunctionException, EvalException {
+    ImmutableMap.Builder<URI, Map<String, List<String>>> headers = new ImmutableMap.Builder<>();
+    for (Map.Entry<String, Map<String, List<String>>> entry : auth.entrySet()) {
+      try {
+        URL url = new URL(entry.getKey());
+        Map<String, List<String>> headerMap = entry.getValue();
+        headers.put(url.toURI(), headerMap);
+      } catch (MalformedURLException e) {
+        throw new RepositoryFunctionException(e, Transience.PERSISTENT);
+      } catch (URISyntaxException e) {
+        throw new EvalException(e);
+      }
+    }
+    return headers.buildOrThrow();
+  }
+
+  private static ImmutableMap<String, Map<String, List<String>>> getHeaderContents(Dict<?, ?> x, String what)
+  throws EvalException {
+    // Dict.cast returns Dict<String, raw Dict>.
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    Map<String, Dict<?, ?>> urlHeaderMapUnchecked = (Map) Dict.cast(x, String.class, Dict.class, what);
+
+    ImmutableMap.Builder<String, Map<String, List<String>>> urlHeadersMap = new ImmutableMap.Builder<>();
+
+    for (Map.Entry<String, Dict<?, ?>> urlHeaderEntry : urlHeaderMapUnchecked.entrySet()) {
+    
+      ImmutableMap.Builder<String, List<String>> headers = new ImmutableMap.Builder<>();
+      Dict<String, List> headersUnchecked = Dict.cast(urlHeaderEntry.getValue(), String.class, List.class, "header entry");
```

Comment:
I think we should `List.cast` every dict value instead of just using `toString()` below.

Reply:
hmm, I don't see List.cast method.